kymatio/scattering2d/backend/numpy_backend.py

```diff
@@ -0,0 +1,176 @@
+        out = np.zeros((input.shape[0], input.shape[1], input.shape[2] // k, input.shape[3] // k), dtype=input.dtype)
+
+        y = input.reshape((input.shape[0], input.shape[1],
+                       input.shape[2]//out.shape[2], out.shape[2],
+                       input.shape[3]//out.shape[3], out.shape[3]))
+
+        out = y.mean(axis=4).mean(axis=2)
+        return out
+
+
+class Modulus(object):
+    """
+        This class implements a modulus transform for complex numbers.
+        Usage
+        -----
+        modulus = Modulus()
+        x_mod = modulus(x)
+        Parameters
+        ---------
+        x: input tensor, with last dimension = 2 for complex numbers
+        Returns
+        -------
+        output: a tensor with imaginary part set to 0, real part set equal to
+        the modulus of x.
+    """
+    def __call__(self, x):
+        norm = np.abs(x)
+        return norm
+
+
+
+
+def fft(input, direction='C2C', inverse=False):
+    """
+        Interface with torch FFT routines for 2D signals.
+        Example
+        -------
+        x = torch.randn(128, 32, 32, 2)
+        x_fft = fft(x, inverse=True)
+        Parameters
+        ----------
+        input : tensor
+            complex input for the FFT
+        direction : string
+            'C2R' for complex to real, 'C2C' for complex to complex
+        inverse : bool
+            True for computing the inverse FFT.
+            NB : if direction is equal to 'C2R', then the transform
+            is automatically inverse.
+    """
+    if direction == 'C2R':
+        if not inverse:
+            raise RuntimeError('C2R mode can only be done with an inverse FFT.')
+
+    if direction == 'C2R':
+        output = np.real(np.fft.ifft2(input, norm=None))*input.shape[-1]*input.shape[-2]
+    elif direction == 'C2C':
+        if inverse:
+            output = np.fft.ifft2(input, norm=None)*input.shape[-1]*input.shape[-2]
+        else:
+            output = np.fft.fft2(input, norm=None)
+
+    return output
+
+
+
+
+def cdgmm(A, B, inplace=False):
+    """
+        Complex pointwise multiplication between (batched) tensor A and tensor B.
+        Parameters
+        ----------
+        A : tensor
+            A is a complex tensor of size (B, C, M, N, 2)
+        B : tensor
+            B is a complex tensor of size (M, N, 2) or real tensor of (M, N, 1)
+        inplace : boolean, optional
+            if set to True, all the operations are performed inplace
+        Returns
+        -------
+        C : tensor
+            output tensor of size (B, C, M, N, 2) such that:
+            C[b, c, m, n, :] = A[b, c, m, n, :] * B[m, n, :]
+    """
+
+    if B.ndim != 2:
+        raise RuntimeError('The filter must be a 3-tensor, with a last '
+                           'dimension of size 1 or 2 to indicate it is real '
+                           'or complex, respectively')
+
+    if inplace:
+        return A.mul_(B)
+    else:
+        return A * B
+
+def new(x, O, M, N):
```

Comment:
yes, cf: https://github.com/edouardoyallon/kymatio/blob/kymatio-dev-frontend/kymatio/scattering2d/core/scattering2d.py#L23 -- I don't know a shared function for tf, np and torch, do you?

Reply:
yeah I don't doubt you are using it. I just find the API a bit weird. it seems super specific. If you can describe it in words, could you put a docstring?

Reply:
I added a very short docstring! a longer one maybe? Actually.. due to the way we code backend, I think it could be also nice to do a sort of "backend template", to avoid changing "backends" times the doc string.... and then to specify it for a specific backend.